asabiris/output/slack/service.py

```diff
@@ -1,38 +1,65 @@
 
-import aiohttp
-
+from slack_sdk import WebClient
+from slack_sdk.errors import SlackApiError
+from typing import List, Tuple
 import asab
 
 from ...output_abc import OutputABC
 
-#
 
 L = logging.getLogger(__name__)
 
-#
-
 
 class SlackOutputService(asab.Service, OutputABC):
-	def __init__(self, app, service_name="SlackOutputService"):
-		super().__init__(app, service_name)
-		# If there is slack configuration section, but no webhook_url, exception is raised.
-		try:
-			self.SlackWebhookUrl = asab.Config.get("slack", "webhook_url")
-		except configparser.NoOptionError as e:
-			L.error("Please provide webhook_url in slack configuration section.")
-			raise e
-		except configparser.NoSectionError:
-			self.SlackWebhookUrl = None
-
-	async def send(self, body):
-		if self.SlackWebhookUrl is None:
-			return
-		async with aiohttp.ClientSession() as session:
-			async with session.post(self.SlackWebhookUrl, json={"text": body}) as resp:
-				if resp.status != 200:
-					L.warning(
-						"Sending alert to Slack was NOT successful. Response status: {}, response: {}".format(
-							resp.status,
-							await resp.text()))
+    def __init__(self, app, service_name="SlackOutputService"):
+        super().__init__(app, service_name)
+        try:
+            self.SlackWebhookUrl = asab.Config.get("slack", "webhook_url")
+            self.Client = WebClient(token=self.SlackWebhookUrl)
+            self.Channel = asab.Config.get("slack", "slack_channel")
+        except configparser.NoOptionError as e:
+            L.error("Please provide webhook_url in slack configuration section.")
+            raise e
+        except configparser.NoSectionError:
+            self.SlackWebhookUrl = None
+
+    async def send(self, body: str, atts: List[Tuple[bytes, str, str]]) -> None:
+        """
+        Sends a message to a Slack channel with optional attachments.
+
+        :param body: The main text of the message to send.
+        :type body: str
+        :param atts: A list of tuples, where each tuple represents an attachment to send. The first item in the tuple is the
+                    attachment's binary data, the second is the attachment's file type, and the third is the attachment's
+                    file name.
+        :type atts: List[Tuple[bytes, str, str]]
+        :return: None
+        :raises SlackApiError: If there was an error sending the message.
+        """
+        if self.Channel is None:
+            L.error("Cannot send message to slack. Reason: Missing slack channel")
+
+        if self.SlackWebhookUrl is None:
+            return
+
+        try:
+            if len(atts) != 0:
+                for attachment in atts:
+                    file_content = attachment[0].encode('utf-8')
+                    file_obj = BytesIO(file_content)
+                    self.Client.files_upload(
+                        channels=self.Channel,
+                        file=file_obj,
+                        initial_comment=body,
```

Comment:
@ateska Is this what you meant by 'Attachments should be part of the Slack message ... not uploaded separately.'?

Reply:
In the single slack message

https://api.slack.com/messaging/webhooks#advanced_message_formatting

See "blocks"